hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenAirdropHandler.java

```diff
@@ -0,0 +1,567 @@
+ */
+@Singleton
+public class TokenAirdropHandler implements TransactionHandler {
+
+    private final CryptoTransferValidator validator;
+
+    /**
+     * Default constructor for injection.
+     */
+    @Inject
+    public TokenAirdropHandler(@NonNull final CryptoTransferValidator validator) {
+        this.validator = validator;
+    }
+
+    @Override
+    public void preHandle(@NonNull final PreHandleContext context) throws PreCheckException {
+        requireNonNull(context);
+
+        final var op = context.body().tokenAirdropOrThrow();
+        final var accountStore = context.createStore(ReadableAccountStore.class);
+        final var tokenStore = context.createStore(ReadableTokenStore.class);
+
+        for (final var transfers : op.tokenTransfers()) {
+            final var tokenID = transfers.tokenOrThrow();
+            final var tokenMeta = tokenStore.getTokenMeta(tokenID);
+            validateTruePreCheck(tokenMeta != null, INVALID_TOKEN_ID);
+            checkFungibleTokenTransfers(tokenID, transfers.transfers(), context, accountStore);
+            checkNftTransfers(tokenID, transfers.nftTransfers(), context, tokenMeta, accountStore);
+        }
+    }
+
+    @Override
+    public void pureChecks(@NonNull final TransactionBody txn) throws PreCheckException {
+        requireNonNull(txn);
+        final var op = txn.tokenAirdropOrThrow();
+        validator.airdropsPureChecks(op);
+    }
+
+    @Override
+    public void handle(@NonNull final HandleContext context) throws HandleException {
+        requireNonNull(context);
+        final var txn = context.body();
+        final var op = txn.tokenAirdropOrThrow();
+        final var pendingStore = context.storeFactory().writableStore(WritableAirdropStore.class);
+        final var accountStore = context.storeFactory().writableStore(WritableAccountStore.class);
+        var recordBuilder = context.recordBuilders().getOrCreate(TokenAirdropRecordBuilder.class);
+        List<TokenTransferList> tokenTransferList = new ArrayList<>();
+
+        // charge custom fees in advance
+        var convertedOp = CryptoTransferTransactionBody.newBuilder()
+                .tokenTransfers(op.tokenTransfers())
+                .build();
+        assessCustomFee(context, convertedOp);
+
+        for (final var xfers : op.tokenTransfers()) {
+            final var tokenId = xfers.tokenOrThrow();
+
+            boolean shouldExecuteCryptoTransfer = false;
+            var transferListBuilder = TokenTransferList.newBuilder().token(tokenId);
+
+            // process fungible token transfers if any
+            if (!xfers.transfers().isEmpty()) {
+                // 1. separate transfers in to two lists
+                // - one list for executing the transfer and one list for adding to pending state
+                var fungibleLists = separateFungibleTransfers(context, tokenId, xfers.transfers());
+                var senderOptionalAmount = xfers.transfers().stream()
```

Comment:
Where do we validate there is exactly one debit in each `tokenTransferList`?

Reply:
+1. That validation should be done in `pureChecks`.

Reply:
Done